libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
+
     SERVE_METRICS_COUNT.inc();
 
-    let mut buffer = vec![];
-    let encoder = TextEncoder::new();
+    // we need to first test this out with vmagent on staging, if it works, we can get turn this to
+    // true or maybe get rid of the fully buffering path altogether
+    const CHUNKED_IS_DEFAULT: bool = false;
+
+    let chunked = super::request::parse_query_param::<_, bool>(&_req, "allow_chunked")?
+        .unwrap_or(CHUNKED_IS_DEFAULT);
+
+    let started_at = std::time::Instant::now();
+
+    if !chunked {
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let mut buffer = Vec::new();
+            let encoder = TextEncoder::new();
+            encoder
+                .encode(&metrics, &mut buffer)
+                .map_err(|e| ApiError::InternalServerError(e.into()))?;
+
+            Ok(Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, encoder.format_type())
+                .body(Body::from(buffer))
+                .unwrap())
+        })
+        .await
+        .map_err(|e| ApiError::InternalServerError(e.into()))
+        .and_then(|e| e)
+    } else {
+        let (tx, rx) = mpsc::channel(1);
+
+        let body = Body::wrap_stream(ReceiverStream::new(rx));
+
+        struct ChannelWriter {
+            buffer: bytes::BytesMut,
+            tx: tokio::sync::mpsc::Sender<std::io::Result<bytes::Bytes>>,
+            written: usize,
+        }
 
-    let metrics = tokio::task::spawn_blocking(move || {
-        // Currently we take a lot of mutexes while collecting metrics, so it's
-        // better to spawn a blocking task to avoid blocking the event loop.
-        metrics::gather()
-    })
-    .await
-    .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
-    encoder.encode(&metrics, &mut buffer).unwrap();
+        impl ChannelWriter {
+            fn new(buf_len: usize, tx: mpsc::Sender<std::io::Result<Bytes>>) -> Self {
+                ChannelWriter {
+                    buffer: BytesMut::with_capacity(buf_len),
+                    tx,
+                    written: 0,
+                }
+            }
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
+                    let ready = self.buffer.split().freeze();
+                    tracing::trace!(n = ready.len(), "flushing");
+
+                    // not ideal to call from blocking code to block_on, but we are sure that this
+                    // operation does not spawn_blocking other tasks
+                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
```

Comment:
```suggestion
                    // not ideal to call from blocking code to block_on, but we are sure that this
                    // operation does not spawn_blocking other tasks, so, no deadlock risk
                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
```

Reply:
Also, do I understand correctly that we use the spawn_blocking thread to block_on the future?

But, further down, we're doing `blocking_send` into the channel if we have a writer error. How does this not deadlock? Because when we're doing the blocking_send, we know we're not writing anymore?

Reply:
This is needed by a `std::io::Write` impl. It is used within the same `spawn_blocking` context as the `blocking_send`, on the same value of `tx` as used here.

So overall this works like:

`Body::wrap_stream(ReceiverStream(rx))` is being fed `std::io::Result<Bytes>` by the `std::io::Write` impl (`ChannelWriter`). The http side of this (stream) is async. The blocking side of this is the prometheus text generation (spawn_blocking).

Do you have a different idea on how to connect the two?

Also I don't understand the deadlock idea at all.

> How does this not deadlock? Because when we're doing the blocking_send, we know we're not writing anymore?

This is ensured by `&mut` which can be thought of as unique borrows here, the slice:

```rust
let mut writer = ChannelWriter::new(128 * 1024, tx);
let encoder = TextEncoder::new();
let res = encoder.encode(&metrics, &mut writer)....;
// the above call produces N of
// block_on(async {
//   self.tx.send(Ok(ready)).await...?;
//   self.tx.reserve().await...?;
// });

// in the Err path:
drop(
    writer
        .tx
        .blocking_send(Err(std::io::ErrorKind::BrokenPipe.into())),
);
```

The `tx` could be a cloned since it's mpsc, but we are just using it as spsc here, because there's no need, and only one producer.